src/main/java/rx/internal/operators/OnSubscribeUsing.java

```diff
@@ -29,38 +33,84 @@
  */
 public final class OnSubscribeUsing<T, Resource> implements OnSubscribe<T> {
 
-    private final Func0<Resource> resourceFactory;
-    private final Func1<? super Resource, ? extends Observable<? extends T>> observableFactory;
-    private final Action1<? super Resource> dispose;
+	private final Func0<Resource> resourceFactory;
+	private final Func1<? super Resource, ? extends Observable<? extends T>> observableFactory;
+	private final Action1<? super Resource> dispose;
+	private final boolean disposeEagerly;
+
+	public OnSubscribeUsing(
+			Func0<Resource> resourceFactory,
+			Func1<? super Resource, ? extends Observable<? extends T>> observableFactory,
+			Action1<? super Resource> dispose, boolean disposeEagerly) {
+		this.resourceFactory = resourceFactory;
+		this.observableFactory = observableFactory;
+		this.dispose = dispose;
+		this.disposeEagerly = disposeEagerly;
+	}
+
+	@Override
+	public void call(Subscriber<? super T> subscriber) {
 
-    public OnSubscribeUsing(Func0<Resource> resourceFactory,
-            Func1<? super Resource, ? extends Observable<? extends T>> observableFactory,
-            Action1<? super Resource> dispose) {
-        this.resourceFactory = resourceFactory;
-        this.observableFactory = observableFactory;
-        this.dispose = dispose;
-    }
+		try {
+			// create the resource
+			final Resource resource = resourceFactory.call();
+			// create an action that disposes only once
+			final Action0 disposeOnceOnly = createOnceOnlyDisposeAction(resource);
```

Comment:
I'd have a class that implements both Action0 and subscription to avoid creating too many objects:
```java
static final class DisposeAction<R> extends AtomicBoolean implements Action0, Subscription {
    Action1<? super R> dispose;
    R resource;
    public DisposeAction(Action1<? super R> dispose, R resource) {
        this.dispose = dispose;
        this.resource = resource;
        lazySet(false); // StoreStore barrier
    }
    @Override
    public void call() {
         if (compareAndSet(false, true)) {
            try {
                dispose.call(resource);
            } finally {
                resource = null;
                dispose = null;
            }
         }
    }
    @Override
    public boolean isUnsubscribed() {
        return get();
    }
    public void unsubscribe() {
        call();
    }
}
```